pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,385 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workspace
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	tenancyinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	tenancylister "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/util/errors"
+)
+
+const targetShardName = "targetShard"
+const unschedulable = "unschedulable"
+
+func NewController(
+	kcpClient kcpclient.ClusterInterface,
+	workspaceInformer tenancyinformer.WorkspaceInformer,
+	workspaceShardInformer tenancyinformer.WorkspaceShardInformer,
+) (*Controller, error) {
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	c := &Controller{
+		queue:                 queue,
+		kcpClient:             kcpClient,
+		workspaceIndexer:      workspaceInformer.Informer().GetIndexer(),
+		workspaceLister:       workspaceInformer.Lister(),
+		workspaceShardIndexer: workspaceShardInformer.Informer().GetIndexer(),
+		workspaceShardLister:  workspaceShardInformer.Lister(),
+	}
+
+	workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+	})
+	if err := c.workspaceIndexer.AddIndexers(map[string]cache.IndexFunc{
+		targetShardName: func(obj interface{}) ([]string, error) {
+			if workspace, ok := obj.(*tenancyv1alpha1.Workspace); ok {
+				return []string{workspace.Spec.TargetShard}, nil
+			}
+			return []string{}, nil
+		},
+		unschedulable: func(obj interface{}) ([]string, error) {
+			if workspace, ok := obj.(*tenancyv1alpha1.Workspace); ok {
+				if IsWorkspaceUnschedulable(workspace) {
+					return []string{"true"}, nil
+				}
+			}
+			return []string{}, nil
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add indexer for Workspace: %w", err)
+	}
+
+	workspaceShardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueAddedShard(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueDeletedShard(obj) },
+	})
+
+	return c, nil
+}
+
+type Controller struct {
```

Comment:
a 2 paragraph descriptions would be nice. Makes it much easier once we have 17 controllers and you lose the overview.

Reply:
It's honestly not even that clear to me what this controller should be doing at the moment or what flows we want/need/can support without more machinery ... hence the confusion.

Reply:
Something like: watches workspaces without spec.target set and tries to find a shard this workspace can be assigned to. The shard then is written to ....